### PR TITLE
Add tests for sagemaker and lambda deployments

### PR DESCRIPTION
**Pull request by yubozhao**

### CODE REVIEW

File: tests/deployment/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import pytest
+
+from bentoml.deployment.utils import ensure_docker_available_or_raise
+from bentoml.exceptions import BentoMLMissingDependencyException, BentoMLException
+
+
+def test_ensure_docker_available_or_raise():
```
Comment (parano) [inline tests/deployment/test_utils.py]: could you explain this test?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 2]
```diff
@@ -31,3 +43,105 @@ def test_sagemaker_handle_client_errors():
...
+
+def test_cleanup_sagemaker_functions():
+    client = boto3.client('sagemaker', 'us-west-2')
+    stubber = Stubber(client)
+    stubber.add_client_error(
+        method='delete_model', service_error_code='ValidationException'
+    )
+    stubber.activate()
+
+    error_status = _cleanup_sagemaker_model(client, 'test_name', 'test_version')
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: test the two cleanup functions separately as unit test, and add integration tests to the entire CreateDeployment implementation, make sure to cover cases like "model creation failed", "endpoint creation failed", "failed to cleanup model", "failed to cleanup endpoint" etc and verify the behavior

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 3]
```diff
@@ -31,3 +43,105 @@ def test_sagemaker_handle_client_errors():
...
+    elif operation_name == 'CreateEndpoint':
+        return {}
+    elif operation_name == 'UpdateEndpoint':
+        return {}
+    return orig(self, operation_name, kwarg)
+
+
+@patch('botocore.client.BaseClient._make_api_call', new=mock_aws_api_calls)
+def test_get_arn_from_aws_user():
+    arn_role = get_arn_role_from_current_aws_user()
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: IIRC get_arn_role_from_current_aws_user has some logic inside, make sure to cover all branches

File: tests/deployment/serverless/test_aws_lambda.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    generate_aws_lambda_handler_py)
+
+
+def test_generate_aws_lambda_serverless_config(tmpdir):
+    python_version = '3.7.2'
+    deployment_name = 'test_deployment_lambda'
+    api_names = ['predict']
+    region = 'us-west-test'
+    namespace = 'namespace'
+    generate_aws_lambda_serverless_config(
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: nit: let's make generate_aws_lambda_serverless_config return the path to the generated config file?

File: bentoml/deployment/sagemaker/__init__.py [hunk 5]
```diff
@@ -165,8 +165,6 @@ def create_push_docker_image_to_ecr(bento_name, bento_version, snapshot_path):
     except ecr_client.exceptions.RepositoryNotFoundException:
         ecr_client.create_repository(repositoryName=image_name)
 
-    if docker_api.tag(image_name, ecr_tag) is False:
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: why was this check here before?

File: bentoml/deployment/sagemaker/__init__.py [hunk 6]
```diff
@@ -261,6 +259,57 @@ def init_sagemaker_project(sagemaker_project_dir, bento_path):
...
+    sagemaker_model_info = {
+        "ModelName": model_name,
+        "PrimaryContainer": {
+            "ContainerHostname": model_name,
+            "Image": ecr_image_path,
+            "Environment": {
+                "API_NAME": api_name,
+                "BENTO_SERVER_TIMEOUT": config().get('apiserver', 'default_timeout'),
+                "BENTO_SERVER_WORKERS": config().get(
+                    'apiserver', 'default_gunicorn_workers_count'
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: this is configured on Yatai server but it's not where the workload is running, let's revisit this later, I believe this should be a user-provided parameter when creating sagemaker deployment, and if it's not there we simply don't set this env var, I believe in that case, we will use the bentoml.server.utils.get_bento_recommend_gunicorn_worker_count to get this value

File: bentoml/deployment/sagemaker/__init__.py [hunk 7]
```diff
@@ -261,6 +259,57 @@ def init_sagemaker_project(sagemaker_project_dir, bento_path):
...
+                ),
+            },
+        },
+        "ExecutionRoleArn": execution_role_arn,
+    }
+
+    logger.info("Creating sagemaker model %s", model_name)
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: can we make these two logging statement the same level? either both info or both debug?

File: bentoml/deployment/sagemaker/__init__.py [hunk 8]
```diff
@@ -261,6 +259,57 @@ def init_sagemaker_project(sagemaker_project_dir, bento_path):
...
+            "ModelName": model_name,
+            "InitialInstanceCount": sagemaker_config.instance_count,
+            "InstanceType": sagemaker_config.instance_type,
+        }
+    ]
+    endpoint_config_name = create_sagemaker_endpoint_config_name(
+        bento_name, bento_version
+    )
+
+    logger.info("Creating Sagemaker endpoint %s configuration", endpoint_config_name)
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: same here

File: tests/deployment/serverless/test_aws_lambda.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    assert yaml_data['provider']['region'] == 'us-west-test'
+
+
+def test_generate_aws_lambda_handler_py(tmpdir):
+    bento_name = 'bento_name'
+    api_names = ['predict', 'second_predict']
+    generate_aws_lambda_handler_py(bento_name, api_names, tmpdir)
+    handler_path = os.path.join(tmpdir, 'handler.py')
+    with open(handler_path, 'r') as f:
+        handler_content = f.read()
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: might be better to load the python module instead of just string matching.

File: bentoml/deployment/sagemaker/__init__.py (hunk 5 above)
Comment (yubozhao) [reply bentoml/deployment/sagemaker/__init__.py]: its not on the latest master, will rebase from that

File: tests/deployment/test_deployment_utils.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+from bentoml.deployment.utils import ensure_docker_available_or_raise
+from bentoml.exceptions import BentoMLMissingDependencyException, BentoMLException
+
+
+def raise_(ex):
+    raise ex
+
+
+def test_ensure_docker_available_or_raise():
```
Comment (parano) [inline tests/deployment/test_deployment_utils.py]: use mock instead of setattr

File: tests/deployment/serverless/test_aws_lambda.py (hunk 9 above)
Comment (yubozhao) [reply tests/deployment/serverless/test_aws_lambda.py]: I would like to load it as module as well. Since this is a fake handler file, the first line `from fake_bento_name import load` will fails.

Maybe you have a good workaround?

File: tests/deployment/test_deployment_utils.py (hunk 10 above)
Comment (yubozhao) [reply tests/deployment/test_deployment_utils.py]: I tried a lot of ways with mock, no luck.

File: tests/deployment/test_utils.py (hunk 1 above)
Comment (yubozhao) [reply tests/deployment/test_utils.py]: file renamed to test_deployment_utils. no longer apply here

File: tests/deployment/sagemaker/test_sagemaker.py (hunk 2 above)
Comment (yubozhao) [reply tests/deployment/sagemaker/test_sagemaker.py]: done

File: tests/deployment/serverless/test_aws_lambda.py (hunk 9 above)
Comment (yubozhao) [reply tests/deployment/serverless/test_aws_lambda.py]: nvm, I can just mock import. lol

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 11]
```diff
@@ -31,3 +44,196 @@ def test_sagemaker_handle_client_errors():
...
+    assert error_status.status_code == Status.UNAUTHENTICATED
+
+    stubber.add_client_error(
+        method='delete_model',
+        service_error_code='RandomError',
+        service_message='random',
+    )
+    with pytest.raises(ClientError) as error:
+        _cleanup_sagemaker_model(client, 'test_name', 'test_version')
+    assert error.value.operation_name == 'DeleteModel'
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: this part looks like just testing Boto3 package?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 12]
```diff
@@ -31,3 +44,196 @@ def test_sagemaker_handle_client_errors():
...
+        if operation_name == 'GetCallerIdentity':
+            return {'Arn': 'something:something:role/random'}
+        elif operation_name == 'GetRole':
+            return {'Role': {'Arn': 'arn:aws:us-west-2:999'}}
+        else:
+            raise Exception(
+                'This test does not handle operation: {}'.format(operation_name)
+            )
+
+    @patch('botocore.client.BaseClient._make_api_call', new=mock_role_path_call)
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]:
```
patch('botocore.client.BaseClient._make_api_call', new=mock_role_path_call)(get_arn_role_from_current_aws_user)
assert get_arn_role_from_current_aws_user() == 'arn:aws:us-west-2:999'
```

Try this?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 13]
```diff
@@ -31,3 +44,196 @@ def test_sagemaker_handle_client_errors():
...
+                                    "Principal": {"Service": "sagemaker.amazonaws.com"},
+                                }
+                            ]
+                        },
+                        "Arn": "arn:aws:us-west-2:888",
+                    }
+                ]
+            }
+        else:
+            raise Exception(
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: no need to, since it's a mock only used here

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 14]
```diff
@@ -31,3 +44,196 @@ def test_sagemaker_handle_client_errors():
     except ClientError as e:
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: change this previous test to use pytest.raises instead of try/except?

File: tests/deployment/sagemaker/test_sagemaker.py (hunk 12 above)
Comment (parano) [reply tests/deployment/sagemaker/test_sagemaker.py]: or 
```
with patch('botocore.client.BaseClient._make_api_call', new=mock_role_path_call): 
    assert get_arn_role_from_current_aws_user() == '....'
```

File: bentoml/deployment/sagemaker/__init__.py [hunk 15]
```diff
@@ -107,7 +107,7 @@ def get_arn_role_from_current_aws_user():
             ):
                 arn = role["Arn"]
         if arn is None:
-            raise ValueError(
+            raise BentoMLDeploymentException(
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: 👍

File: bentoml/deployment/utils.py [hunk 16]
```diff
@@ -57,52 +53,20 @@ def process_docker_api_line(payload):
...
-    # copy the generated targz to archive directory and remove it from
-    # bentoml module directory
-    dest_dir = os.path.join(archive_path, 'bundled_pip_dependencies')
-    shutil.copytree(source_dir, dest_dir)
-    shutil.rmtree(source_dir)
-
-
 def ensure_docker_available_or_raise():
+    # for FileNotFoundError doesn't exist in py2.7. check_output raise OSError instead
+    if version_info.major < 3:
```
Comment (parano) [inline bentoml/deployment/utils.py]: use ```if six.PY3:``` ?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 17]
```diff
@@ -1,33 +1,419 @@
...
+@pytest.fixture(scope='function')
+def sagemaker_client():
+    return boto3.client('sagemaker', 'us-west-2')
 
-def test_sagemaker_handle_client_errors():
-    client = boto3.client('sagemaker', 'us-west-2')
-    stubber = Stubber(client)
+
+def test_sagemaker_handle_client_errors(sagemaker_client):
+    sagemaker_client = boto3.client('sagemaker', 'us-west-2')
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: remove this line?

Comment (parano) [reply tests/deployment/sagemaker/test_sagemaker.py]: since you have the pytest fixture now

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 18]
```diff
@@ -1,33 +1,419 @@
...
-    except ClientError as e:
-        result = _parse_aws_client_exception_or_raise(e)
+    with pytest.raises(ClientError) as e:
+        sagemaker_client.describe_endpoint(EndpointName='Test')
+    result = _parse_aws_client_exception_or_raise(e.value)
     assert result.status_code == Status.UNAUTHENTICATED
+
+
+def test_cleanup_sagemaker_model(sagemaker_client):
+    sagemaker_client = boto3.client('sagemaker', 'us-west-2')
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: remove?

Comment (parano) [reply tests/deployment/sagemaker/test_sagemaker.py]: or just remove the fixture, I think either way is fine

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 19]
```diff
@@ -1,33 +1,419 @@
...
+    stubber = Stubber(sagemaker_client)
+    stubber.add_client_error(
+        method='delete_model', service_error_code='ValidationException'
+    )
+    stubber.activate()
+
+    error_status = _cleanup_sagemaker_model(
+        sagemaker_client, 'test_name', 'test_version'
+    )
+    assert error_status.status_code == Status.NOT_FOUND
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: probably a follow up issue, why is ValidationException mapping to NOT_FOUND? 

I saw the code in `_parse_aws_client_exception_or_raise`:

```
    if error_code == 'ValidationException':
        logger.error(error_log_message)
        return Status.NOT_FOUND(error_response.get('Message', 'Unknown'))
```
Shouldn't it also check the actual error_message? There might be other cases when AWS returns a ValidationException right?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 20]
```diff
@@ -1,33 +1,419 @@
...
 
+@pytest.fixture(scope='function')
+def sagemaker_client():
+    return boto3.client('sagemaker', 'us-west-2')
 
-def test_sagemaker_handle_client_errors():
-    client = boto3.client('sagemaker', 'us-west-2')
-    stubber = Stubber(client)
+
+def test_sagemaker_handle_client_errors(sagemaker_client):
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: this is enssentially a unit test for function ```_parse_aws_client_exception_or_raise```, make it simpler without the mock staff:
```
err = ClientError(
           error_response={
               'Error': {'Code': 'MyCode', 'Message': 'MyMessage'}},
           operation_name='myoperation'
         )
result = _parse_aws_client_exception_or_raise(err)
assert result....

```

File: bentoml/deployment/sagemaker/__init__.py [hunk 21]
```diff
@@ -206,14 +207,12 @@ def _parse_aws_client_exception_or_raise(e):
...
-        logger.error(error_log_message)
-        return Status.NOT_FOUND(error_response.get('Message', 'Unknown'))
+        return Status.NOT_FOUND(error_log_message)
     elif error_code == 'InvalidSignatureException':
-        logger.error(error_log_message)
-        return Status.UNAUTHENTICATED(error_response.get('Message', 'Unknown'))
+        return Status.UNAUTHENTICATED(error_log_message)
     else:
-        logger.error(error_log_message)
         raise e
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: Realized an issue when reading through the tests, here it should raise an Exception that is inherited from BentoMLException, only that way the RPC server can return a response with INTERNAL ERROR code I believe

```
raise BentoMLException(
            'Error interacting with AWS service: {}'.format(str(e))
        )
```

File: bentoml/deployment/sagemaker/__init__.py [hunk 22]
```diff
@@ -259,6 +258,57 @@ def init_sagemaker_project(sagemaker_project_dir, bento_path):
...
+):
+    production_variants = [
+        {
+            "VariantName": generate_aws_compatible_string(bento_name),
+            "ModelName": model_name,
+            "InitialInstanceCount": sagemaker_config.instance_count,
+            "InstanceType": sagemaker_config.instance_type,
+        }
+    ]
+    endpoint_config_name = create_sagemaker_endpoint_config_name(
```
Comment (parano) [inline bentoml/deployment/sagemaker/__init__.py]: can we rename those two functions to verb `get` and mark as private in the module? for example:
```
def _get_sagemaker_model_name(bento_name, bento_version):
   ...

def _get_sagemaker_endpoint_config_name(bento_name, bento_version)
   ...
```
or 
```
def _sagemaker_model_name(bento_name, bento_version):
   ...

def _sagemaker_endpoint_config_name(bento_name, bento_version)
   ...
```

File: bentoml/deployment/serverless/serverless_utils.py [hunk 23]
```diff
@@ -40,7 +39,9 @@
 SERVERLESS_BIN_COMMAND = '{}/node_modules/.bin/serverless'.format(BENTOML_HOME)
 
 
-def check_nodejs_comptaible_version():
+def check_nodejs_compatible_version():
+    from bentoml.utils.whichcraft import which
```
Comment (parano) [inline bentoml/deployment/serverless/serverless_utils.py]: could you test if the tests pass now without the change? would love to understand why this change was made

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 24]
```diff
@@ -1,33 +1,419 @@
...
+
+
+def test_cleanup_sagemaker_endpoint_config(sagemaker_client):
+    stubber = Stubber(sagemaker_client)
+    stubber.add_client_error(
+        method='delete_endpoint_config', service_error_code='ValidationException'
+    )
+    stubber.activate()
+
+    error_status = _cleanup_sagemaker_endpoint_config(
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: ```_cleanup_sagemaker_endpoint_config``` really should not be returning an error status, will definitely need to do some refactoring there, could you just remove these two tests `test_cleanup_sagemaker_model` and `test_cleanup_sagemaker_endpoint_config`  for now? they are just testing one line logic anyway

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 25]
```diff
@@ -1,33 +1,419 @@
...
+            return {'Arn': 'something:something:role/random'}
+        elif operation_name == 'GetRole':
+            return {'Role': {'Arn': ROLE_PATH_ARN_RESULT}}
+        else:
+            raise Exception(
+                'This test does not handle operation: {}'.format(operation_name)
+            )
+
+    @patch('botocore.client.BaseClient._make_api_call', new=mock_role_path_call)
+    def role_path():
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: use the following syntax  instead of wrapping function:

```
with patch('botocore.client.BaseClient._make_api_call', new=mock_role_path_call): 
    assert get_arn_role_from_current_aws_user() == '....'
```

File: tests/deployment/serverless/test_aws_lambda.py [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+def test_generate_aws_lambda_handler_py(tmpdir):
+    bento_name = 'bento_name'
+    api_names = ['predict', 'second_predict']
+    generate_aws_lambda_handler_py(bento_name, api_names, str(tmpdir))
+
+    import sys
+
+    sys.modules['bento_name'] = Mock()
+
+    sys.path.insert(1, str(tmpdir))
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: remove it from sys.path afterwards

File: tests/deployment/serverless/test_aws_lambda.py [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+
+
+def test_generate_aws_lambda_handler_py(tmpdir):
+    bento_name = 'bento_name'
+    api_names = ['predict', 'second_predict']
+    generate_aws_lambda_handler_py(bento_name, api_names, str(tmpdir))
+
+    import sys
+
+    sys.modules['bento_name'] = Mock()
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: remove it after the test?

Comment (parano) [reply tests/deployment/serverless/test_aws_lambda.py]: and would be nice to verify `Mock.load` is called, and have it return  an object with `get_service_api('predict')`, that way you can actually call the `predict` api function imported below

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 28]
```diff
@@ -1,33 +1,419 @@
...
+    @patch(mock_open_param_value, mock_open(read_data='test'), create=True)
+    @patch('shutil.copytree', autospec=True)
+    @patch('os.chmod', autospec=True)
+    def mock_wrapper(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    return mock_wrapper
+
+
+def create_yatai_service(is_local_repo=True):
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: rename: create_yatai_service_mock?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 29]
```diff
@@ -1,33 +1,419 @@
...
+    yatai_service = create_yatai_service()
+    sagemaker_deployment_pb = generate_sagemaker_deployment_pb()
+    deployment_operator = SageMakerDeploymentOperator()
+    result_pb = deployment_operator.apply(sagemaker_deployment_pb, yatai_service)
+    assert result_pb.status.status_code == Status.OK
+    assert result_pb.deployment.name == TEST_DEPLOYMENT_NAME
+
+
+# @mock_sagemaker_deployment_wrapper
+# def test_sagemaker_apply_model_already_exists(
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: why this one was commented out?

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 30]
```diff
@@ -1,33 +1,419 @@
...
+    test_deployment_pb.spec.sagemaker_operator_config.api_name = TEST_BENTO_API_NAME
+    test_deployment_pb.spec.sagemaker_operator_config.region = TEST_AWS_REGION
+    test_deployment_pb.spec.sagemaker_operator_config.instance_count = (
+        TEST_DEPLOYMENT_INSTANCE_COUNT
+    )
+    test_deployment_pb.spec.sagemaker_operator_config.instance_type = (
+        TEST_DEPLOYMENT_INSTANCE_TYPE
+    )
+    # DeploymentSpec.DeploymentOperator.AWS_SAGEMAKER
+    test_deployment_pb.spec.operator = 2
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: just use
```
test_deployment_pb.spec.operator = deployment_pb2.DeploymentSpec.AWS_SAGEMAKER
```

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 31]
```diff
@@ -1,33 +1,419 @@
...
+
+
+def create_yatai_service(is_local_repo=True):
+    def mock_get_bento(is_local_repo=True):
+        bento_pb = Bento(
+            name=TEST_DEPLOYMENT_BENTO_NAME, version=TEST_DEPLOYMENT_BENTO_VERSION
+        )
+        # BentoUri.StorageType.LOCAL
+        if is_local_repo:
+            bento_pb.uri.type = 1
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: similarily, use ```bento_pb.uri.type = repository_pb2.BentoUri.LOCAL``` ?

File: tests/deployment/sagemaker/test_sagemaker.py (hunk 28 above)
Comment (parano) [reply tests/deployment/sagemaker/test_sagemaker.py]: `is_local` was a big confusing, consider use a parameter like bento_storage_type?
```create_yatai_service_mock(bento_storage_type=BentoUri.LOCAL)```

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 32]
```diff
@@ -1,33 +1,419 @@
...
+    return test_deployment_pb
+
+
+def raise_(ex):
+    raise ex
+
+
+@mock_sagemaker_deployment_wrapper
+def test_sagemaker_apply_fail_not_local_repo(
+    mock_chmod, mock_copytree, mock_docker_push, mock_docker_build, mock_check_output
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: since those mocks are referenced here, could we just do a sanity check on the ones that actually expected to get called and omit the ones that are not needed? same for the other a few tests in this file

File: tests/deployment/serverless/test_serverless_utils.py [hunk 33]
```diff
@@ -0,0 +1,14 @@
...
+
+from bentoml.deployment.serverless.serverless_utils import (
+    check_nodejs_compatible_version,
+)
+from bentoml.exceptions import BentoMLMissingDependencyException
+
+
+def test_check_nodejs_compatible_version():
+    with pytest.raises(BentoMLMissingDependencyException) as error:
+        with patch('bentoml.utils.whichcraft.which', return_value=None):
```
Comment (parano) [inline tests/deployment/serverless/test_serverless_utils.py]: Oh I think I know why that 'import' change was made now.. ignore that previous comment

Comment (parano) [reply tests/deployment/serverless/test_serverless_utils.py]: Would be great to add a comment at that function tho

File: tests/deployment/serverless/test_aws_lambda.py [hunk 34]
```diff
@@ -0,0 +1,194 @@
...
+    @patch.object(serverless_utils, 'init_serverless_project_dir', autospec=True)
+    @patch('subprocess.Popen')
+    @patch('botocore.client.BaseClient._make_api_call', new=mock_describe_cf)
+    def wrapper(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def generate_fake_yatai_service(is_local=False):
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: nit:  rename to `generate_yatai_service_mock`

File: tests/deployment/serverless/test_aws_lambda.py [hunk 35]
```diff
@@ -0,0 +1,194 @@
...
+    mock_popen,
+    mock_init_serverless,
+    mock_copy,
+    mock_copytree,
+    mock_check_nodejs,
+    mock_checkcall,
+    mock_checkoutput,
+):
+    test_deployment_pb = generate_lambda_deployment_pb()
+    fake_yatai_service = generate_fake_yatai_service()
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: nit: s/fake_yatai_service/yatai_service_mock

File: tests/deployment/serverless/test_aws_lambda.py (hunk 34 above)
Comment (parano) [reply tests/deployment/serverless/test_aws_lambda.py]: and about is_local parameter, see previous comment

Also since majority of the tests are using local repository, would be nice to use that as default instead of making non-local a default here

File: tests/deployment/sagemaker/test_sagemaker.py [hunk 36]
```diff
@@ -1,33 +1,338 @@
...
+    @patch('docker.APIClient.build', autospec=True)
+    @patch('docker.APIClient.push', autospec=True)
+    @patch('bentoml.deployment.sagemaker.init_sagemaker_project', autospec=True)
+    def mock_wrapper(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    return mock_wrapper
+
+
+def create_yatai_service_mock(repo_storage_type=BentoUri.LOCAL):
```
Comment (parano) [inline tests/deployment/sagemaker/test_sagemaker.py]: recommend cleanup:
```
def create_yatai_service_mock(repo_storage_type=BentoUri.LOCAL):
        bento_pb = Bento(
            name=TEST_DEPLOYMENT_BENTO_NAME, version=TEST_DEPLOYMENT_BENTO_VERSION
        )
        if repo_storage_type == BentoUri.LOCAL:
            bento_pb.uri.uri = TEST_DEPLOYMENT_BENTO_LOCAL_URI
        bento_pb.uri.type = repo_storage_type
        api = BentoServiceMetadata.BentoServiceApi(name=TEST_BENTO_API_NAME)
        bento_pb.bento_service_metadata.apis.extend([api])
        get_bento_response = GetBentoResponse(bento=bento_pb)

       yatai_service_mock = MagicMock()
       yatai_service_mock.GetBento.return_value = get_bento_response

      return yatai_service_mock
```

File: tests/deployment/serverless/test_aws_lambda.py [hunk 37]
```diff
@@ -0,0 +1,212 @@
...
+    @patch.object(serverless_utils, 'init_serverless_project_dir', autospec=True)
+    @patch('subprocess.Popen')
+    @patch('botocore.client.BaseClient._make_api_call', new=mock_describe_cf)
+    def wrapper(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def create_yatai_service_mock(repo_storage_type=BentoUri.LOCAL):
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]: similarily as the previous comment

File: tests/deployment/serverless/test_aws_lambda.py [hunk 38]
```diff
@@ -0,0 +1,212 @@
...
+            raise Exception('This test does not handle operation {}'.format(op_name))
+
+    @patch('botocore.client.BaseClient._make_api_call', new=mock_cf_response)
+    def mock_describe(deployment_pb, yatai_service):
+        deployment_operator = AwsLambdaDeploymentOperator()
+        return deployment_operator.describe(deployment_pb, yatai_service)
+
+    yatai_service_mock = create_yatai_service_mock()
+    test_deployment_pb = generate_lambda_deployment_pb()
+    result_pb = mock_describe(test_deployment_pb, yatai_service_mock)
```
Comment (parano) [inline tests/deployment/serverless/test_aws_lambda.py]:
```
with patch('botocore.client.BaseClient._make_api_call', new=mock_cf_response):
        deployment_operator = AwsLambdaDeploymentOperator()
        result_pb  = deployment_operator.describe(test_deployment_pb, yatai_service_mock)
        assert result_pb.status.status_code == Status.OK
```
